<?xml version="1.0"?>

<!-- Adapted for 3DX use by Numurus, Copyright (c) 2021 -->

<!--
Copyright (c) 2020, STEREOLABS.

All rights reserved.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-->

<robot name="numurus_3dx_c" xmlns:xacro="http://ros.org/wiki/xacro">

<!-- Includes -->
    <xacro:include filename="$(find nepi_ros_interfaces)/urdf/include/materials.urdf.xacro" />

    <!-- Arguments -->
    <xacro:arg name="camera_name"   default="stereo_cam" />
    <xacro:arg name="camera_model"  default="zedm" />
    <xacro:arg name="base_frame"    default="base_link" />
    <xacro:arg name="cam_pos_x"     default="-0.0294" />
    <xacro:arg name="cam_pos_y"     default="0.0" />
    <xacro:arg name="cam_pos_z"     default="0.0" />
    <xacro:arg name="cam_roll"      default="0.0" />
    <xacro:arg name="cam_pitch"     default="0.0" />
    <xacro:arg name="cam_yaw"       default="0.0" />

    <!-- Properties -->
    <xacro:property name="model"    value="$(arg camera_model)" />
    <xacro:property name="M_PI"     value="3.1415926535897931" />

    <xacro:if value="${model == 'zed'}">
        <xacro:property name="baseline" value="0.12" />
    </xacro:if>
    <xacro:if value="${model == 'zedm'}">
        <xacro:property name="baseline" value="0.06" />
    </xacro:if>
    <xacro:if value="${model == 'zed2'}">
        <xacro:property name="baseline" value="0.12" />
    </xacro:if>

<!-- base_link -->
    <link name="$(arg base_frame)" />

<!-- pan/tilt TODO -->

<!-- 3DX Center Joint -->
    <joint name="3dx_center_joint" type="fixed">
        <parent link="$(arg base_frame)"/>
        <child link="3dx_center_frame"/>
        <origin xyz="0 0 0" rpy="0 0 0" />
    </joint>

    <link name="3dx_center_frame">
      <visual>
        <origin xyz="-0.15 0.06 0.026" rpy="-1.57075 0 -1.57075"/>
        <geometry>
          <mesh filename="package://nepi_ros_interfaces/urdf/models/3dx_c.stl" name="3dx_mat" scale="0.001 0.001 0.001"/>
        </geometry>
        <material name="3dx_mat" />
      </visual>
    </link>

<!-- Camera Center -->
    <joint name="$(arg camera_name)_camera_center_joint" type="fixed">
        <parent link="3dx_center_frame"/>
        <child link="$(arg camera_name)_camera_center"/>
        <origin xyz="$(arg cam_pos_x) $(arg cam_pos_y) $(arg cam_pos_z)" rpy="$(arg cam_roll) $(arg cam_pitch) $(arg cam_yaw)" />
    </joint>

    <link name="$(arg camera_name)_camera_center"/>

<!-- Left Camera -->
    <joint name="$(arg camera_name)_left_camera_joint" type="fixed">
        <parent link="$(arg camera_name)_camera_center"/>
        <child link="$(arg camera_name)_left_camera_frame"/>
        <origin xyz="0 ${baseline/2} 0" rpy="0 0 0" />
    </joint>

    <link name="$(arg camera_name)_left_camera_frame" />

    <joint name="$(arg camera_name)_left_camera_optical_joint" type="fixed">
        <origin xyz="0 0 0" rpy="-${M_PI/2} 0.0 -${M_PI/2}"/>
        <parent link="$(arg camera_name)_left_camera_frame"/>
        <child link="$(arg camera_name)_left_camera_optical_frame"/>
    </joint>

    <link name="$(arg camera_name)_left_camera_optical_frame"/>

<!-- Right Camera -->
    <joint name="$(arg camera_name)_right_camera_joint" type="fixed">
        <parent link="$(arg camera_name)_camera_center"/>
        <child link="$(arg camera_name)_right_camera_frame"/>
        <origin xyz="0 -${baseline/2} 0" rpy="0 0 0" />
    </joint>

    <link name="$(arg camera_name)_right_camera_frame" />

    <joint name="$(arg camera_name)_right_camera_optical_joint" type="fixed">
        <origin xyz="0 0 0" rpy="-${M_PI/2} 0.0 -${M_PI/2}"/>
        <parent link="$(arg camera_name)_right_camera_frame"/>
        <child link="$(arg camera_name)_right_camera_optical_frame"/>
    </joint>

    <link name="$(arg camera_name)_right_camera_optical_frame"/>

<!-- ZED2 Sensors -->
    <xacro:if value="${model == 'zed2'}">
        <joint name="$(arg camera_name)_mag_joint" type="fixed">
            <parent link="$(arg camera_name)_camera_center"/>
            <child link="$(arg camera_name)_mag_link"/>
            <origin xyz="0 0 0" rpy="0 0 0" />
        </joint>

        <link name="$(arg camera_name)_mag_link" />

        <joint name="$(arg camera_name)_baro_joint" type="fixed">
            <parent link="$(arg camera_name)_camera_center"/>
            <child link="$(arg camera_name)_baro_link"/>
            <origin xyz="0 0 0" rpy="0 0 0" />
        </joint>

        <link name="$(arg camera_name)_baro_link" />

        <joint name="$(arg camera_name)_temp_left_joint" type="fixed">
            <parent link="$(arg camera_name)_left_camera_frame"/>
            <child link="$(arg camera_name)_temp_left_link"/>
            <origin xyz="0 0 0" rpy="0 0 0" />
        </joint>

        <link name="$(arg camera_name)_temp_left_link" />

        <joint name="$(arg camera_name)_temp_right_joint" type="fixed">
            <parent link="$(arg camera_name)_right_camera_frame"/>
            <child link="$(arg camera_name)_temp_right_link"/>
            <origin xyz="0 0 0" rpy="0 0 0" />
        </joint>

        <link name="$(arg camera_name)_temp_right_link" />
    </xacro:if>
</robot>
